src/CFamily.UnitTests/CMake/CompilationConfigProviderTests.cs

```diff
@@ -0,0 +1,236 @@
+
+            act.Should().Throw<StackOverflowException>();
+        }
+
+        [TestMethod]
+        public void GetConfig_ProblemParsingDatabaseFile_Null()
+        {
+            var compilationDatabaseLocator = SetupCompilationDatabaseLocator("some db");
+            var fileSystem = SetupDatabaseFileContents("some db", "not valid json");
+            var logger = new TestLogger();
+
+            var testSubject = CreateTestSubject(compilationDatabaseLocator.Object, fileSystem.Object, logger);
+
+            var result = testSubject.GetConfig("some file");
+            result.Should().BeNull();
+
+            fileSystem.Verify(x => x.File.ReadAllText("some db"), Times.Once);
+
+            logger.AssertPartialOutputStringExists("JsonReaderException");
+        }
+
+        [TestMethod]
+        [DataRow("")]
+        [DataRow("[]")]
+        [DataRow("[{}]")]
+        [DataRow("[{\"file\" : \"some file.c\" }]")] // different extension
+        [DataRow("[{\"file\" : \"sub/some file.cpp\" }]")] // different path
+        public void GetConfig_EntryIsNotFoundInDatabase_Null(string databaseFileContents)
+        {
+            var compilationDatabaseLocator = SetupCompilationDatabaseLocator("some db");
+            var fileSystem = SetupDatabaseFileContents("some db", databaseFileContents);
+            var logger = new TestLogger();
+
+            var testSubject = CreateTestSubject(compilationDatabaseLocator.Object, fileSystem.Object, logger);
+
+            var result = testSubject.GetConfig("some file.cpp");
+            result.Should().BeNull();
+
+            fileSystem.Verify(x => x.File.ReadAllText("some db"), Times.Once);
+
+            logger.AssertPartialOutputStringExists(string.Format(Resources.NoCompilationDatabaseEntry, "some file.cpp"));
+        }
+
+        [TestMethod]
+        [DataRow("c:\\some file.cpp")] // exact match
+        [DataRow("c:/some file.cpp")] // different format 
+        [DataRow("c:/SOME file.cpp")] // case-sensitivity on name
+        [DataRow("c:/some file.CPP")] // case-sensitivity on extension
+        public void GetConfig_EntryFound_ReturnsEntry(string entryFilePath)
+        {
+            var compilationDatabaseLocator = SetupCompilationDatabaseLocator("some db");
+            var compilationDatabaseContent = new[]
+            {
+                new CompilationDatabaseEntry {File = entryFilePath, Command = "some command", Directory = "some dir"}
+            };
+            var fileSystem = SetupDatabaseFileContents("some db", JsonConvert.SerializeObject(compilationDatabaseContent));
+            var logger = new TestLogger();
+
+            var testSubject = CreateTestSubject(compilationDatabaseLocator.Object, fileSystem.Object, logger);
+
+            var result = testSubject.GetConfig("c:\\some file.cpp");
+            result.Should().BeEquivalentTo(compilationDatabaseContent[0]);
+        }
+
+        [TestMethod]
+        public void GetConfig_MultipleEntriesForFile_ReturnsFirstOne()
```

Comment:
Is this a valid scenario, or is the produce code just being defensive?

If the product code is just being defensive then it would be worth adding a comment to make it clear e.g. 
`Defensive: a valid compilation database should only have at most one entry per file`.

If a compilation database can validly have multiple entries for the same file then we probably need to do more work on our side.

Reply:
It can be included multiple times, yes, for instance if it's referenced by two different projects.
I don't know how we would go about getting the right compilation in this case, I'll ask Abbas

Reply:
FYI according to Abbas it's possible but unlikely, so any entry will do.